File: Example/CoreData/SyncKit/QSSharedCompanyTableViewController.m
//
//  QSSharedCompanyTableViewController.m
//  SyncKitCoreDataExample
//
//  Created by Manuel Entrena on 20/04/2018.
//  Copyright © 2018 Manuel. All rights reserved.
//

#import "QSSharedCompanyTableViewController.h"
#import "QSCompany.h"
#import "QSEmployeeTableViewController.h"
#import "QSCompanyTableViewCell.h"
@import SyncKit;

@interface QSSharedCompanyTableViewController () <QSCoreDataMultiFetchedResultsControllerDelegate, UICloudSharingControllerDelegate>

@property (nonatomic, weak) IBOutlet UIButton *syncButton;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *indicatorView;

@property (nonatomic, strong) QSCompany *sharingCompany;

@property (nonatomic, strong) QSCoreDataMultiFetchedResultsController *fetchedResultsController;

@property (nonatomic, weak) IBOutlet UIView *loadingView;

@end

@implementation QSSharedCompanyTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupFetchedResultsController];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setupFetchedResultsController
{
    if (!_fetchedResultsController) {
        NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"QSCompany"];
        fetchRequest.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES]];
        
        _fetchedResultsController = [self.synchronizer multiFetchedResultsControllerWithFetchRequest:fetchRequest];
        _fetchedResultsController.delegate = self;
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"showEmployees"]) {
        QSEmployeeTableViewController *employeeTableViewController = (QSEmployeeTableViewController *)segue.destinationViewController;
        QSCompany *company = (QSCompany *)sender;
        employeeTableViewController.company = company;
        employeeTableViewController.managedObjectContext = employeeTableViewController.company.managedObjectContext;
        CKShare *share = [self.synchronizer shareFor:company];
        employeeTableViewController.canWrite = share.currentUserParticipant.permission == CKShareParticipantPermissionReadWrite;
    }
}

#pragma mark - NSFetchedResultsControllerDelegate

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView beginUpdates];
}


- (void)controller:(NSFetchedResultsController *)controller didChangeSection:(id <NSFetchedResultsSectionInfo>)sectionInfo
           atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type {
    
    switch(type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex]
                          withRowAnimation:UITableViewRowAnimationFade];
            break;
        default:
            break;
    }
}


- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath {
    
    UITableView *tableView = self.tableView;
    
    switch(type) {
            
        case NSFetchedResultsChangeInsert:
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeDelete:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
            
        case NSFetchedResultsChangeUpdate:
            [self configureCell:[tableView cellForRowAtIndexPath:indexPath]
                    atIndexPath:indexPath];
            break;
            
        case NSFetchedResultsChangeMove:
            [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]
                             withRowAnimation:UITableViewRowAnimationFade];
            break;
    }
}


- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller {
    [self.tableView endUpdates];
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    QSCompany *company = [self objectAtIndexPath:indexPath];
    [self performSegueWithIdentifier:@"showEmployees" sender:company];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    
    return self.fetchedResultsController.fetchedResultsControllers.count;
}

- (NSInteger)tableView:(UITableView *)table numberOfRowsInSection:(NSInteger)section {
    
    return self.fetchedResultsController.fetchedResultsControllers[section].fetchedObjects.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    QSCompanyTableViewCell *cell = (QSCompanyTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"cell"];
    [self configureCell:cell atIndexPath:indexPath];
    
    return cell;
}

- (void)configureCell:(QSCompanyTableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath
{
    QSCompany *company = [self objectAtIndexPath:indexPath];
    cell.nameLabel.text = company.name;
    [cell.sharingButton setTitle:@"Shared with me" forState:UIControlStateNormal];
    
    __weak QSSharedCompanyTableViewController *weakSelf = self;
    
    cell.shareButtonAction = ^{
        [weakSelf shareCompany:company];
    };
}

- (QSCompany *)objectAtIndexPath:(NSIndexPath *)indexPath
{
    return self.fetchedResultsController.fetchedResultsControllers[indexPath.section].fetchedObjects[indexPath.row];
}

#pragma mark - Actions

- (IBAction)didTapSynchronize:(id)sender
{
    [self synchronizeWithCompletion:nil];
}

- (IBAction)synchronizeWithCompletion:(void(^)(NSError *error))completion
{
    [self showLoading:YES];
    __weak QSSharedCompanyTableViewController *weakSelf = self;
    [self.synchronizer synchronizeWithCompletion:^(NSError *error) {
        [weakSelf showLoading:NO];
        if (error) {
            UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:[NSString stringWithFormat:@"Error: %@", error] preferredStyle:UIAlertControllerStyleAlert];
            
            [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
            
            [weakSelf presentViewController:alertController animated:YES completion:nil];
        }
        if (completion) {
            completion(error);
        }
    }];
}

- (void)shareCompany:(QSCompany *)company
{
    self.sharingCompany = company;
    
    [self showLoading:YES];
    __weak QSSharedCompanyTableViewController *weakSelf = self;
    
    [self.synchronizer synchronizeWithCompletion:^(NSError *error) {
        
        [weakSelf showLoading:NO];
        if (!error) {
            UICloudSharingController *sharingController;
            CKShare *share = [weakSelf.synchronizer shareFor:company];
            CKContainer *container = [CKContainer containerWithIdentifier:weakSelf.synchronizer.containerIdentifier];
            if (share) {
                sharingController = [[UICloudSharingController alloc] initWithShare:share container:container];
            } else {
                sharingController = [[UICloudSharingController alloc] initWithPreparationHandler:^(UICloudSharingController * _Nonnull controller, void (^ _Nonnull preparationCompletionHandler)(CKShare * _Nullable, CKContainer * _Nullable, NSError * _Nullable)) {
                    
                    [self.synchronizer shareWithObject:company publicPermission:CKShareParticipantPermissionReadOnly participants:@[] completion:^(CKShare *share, NSError *error) {
                        
                        preparationCompletionHandler(share, container, error);
                    }];
                }];
            }
            
            sharingController.availablePermissions = UICloudSharingPermissionAllowPublic | UICloudSharingPermissionAllowReadOnly | UICloudSharingPermissionAllowReadWrite;
            sharingController.delegate = self;
            [weakSelf presentViewController:sharingController animated:YES completion:nil];
        }
    }];
}

#pragma mark - UICloudSharingControllerDelegate

- (NSString *)itemTitleForCloudSharingController:(UICloudSharingController *)csc
{
    return self.sharingCompany.name;
}

- (void)cloudSharingController:(UICloudSharingController *)csc failedToSaveShareWithError:(NSError *)error
{
    NSLog(@"failed to save share");
}

- (void)cloudSharingControllerDidSaveShare:(UICloudSharingController *)csc
{
    [self.synchronizer saveShare:csc.share for:self.sharingCompany];
    [self.synchronizer synchronizeWithCompletion:nil];
}

- (void)cloudSharingControllerDidStopSharing:(UICloudSharingController *)csc
{
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self synchronizeWithCompletion:nil];
    });
}

- (void)multiFetchedResultsControllerDidChangeControllers:(QSCoreDataMultiFetchedResultsController *)controller {
    
    [self.tableView reloadData];
}

#pragma mark - Loading

- (void)showLoading:(BOOL)loading
{
    if (loading) {
        self.syncButton.hidden = YES;
        self.loadingView.hidden = NO;
        [self.indicatorView startAnimating];
    } else {
        self.syncButton.hidden = NO;
        self.loadingView.hidden = YES;
        [self.indicatorView stopAnimating];
    }
}

@end

File: Example/CoreData/SyncKit/QSSharedCompanyTableViewController.swift
//
//  QSSharedCompanySwiftTableViewController.swift
//  SyncKitCoreDataExample
//
//  Created by Jérôme Haegeli on 25.07.18.
//  Copyright © 2018 Manuel. All rights reserved.
//

import UIKit
import SyncKit

class QSSharedCompanyTableViewController: UITableViewController, QSCoreDataMultiFetchedResultsControllerDelegate, UICloudSharingControllerDelegate {
    
    var synchronizer: QSCloudKitSynchronizer?
    
    @IBOutlet weak var syncButton: UIButton!
    @IBOutlet weak var indicatorView: UIActivityIndicatorView!
    
    var sharingCompany: QSCompany!
    var fetchedResultsController: QSCoreDataMultiFetchedResultsController?
    
    @IBOutlet weak var loadingView: UIView?

    
    override func viewDidLoad() {
        super.viewDidLoad()
        subscribeChanges()
        self.setupFetchedResultsController()
    
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        self.tableView.reloadData()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    
    func setupFetchedResultsController() {
        if !(fetchedResultsController != nil) {
            let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "QSCompany")
            fetchRequest.sortDescriptors = [NSSortDescriptor(key: "name", ascending: true)]
            fetchedResultsController = synchronizer?.multiFetchedResultsController(with: fetchRequest)
            fetchedResultsController?.delegate = self
            print(fetchedResultsController)
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if (segue.identifier == "showEmployees") {
            guard let employeeTableViewController = segue.destination as? QSEmployeeTableViewController else {
                fatalError("Application storyboard mis-configuration")
            }
            
            let company = sender as? QSCompany
            employeeTableViewController.company = company

            if let aContext = employeeTableViewController.company?.managedObjectContext {
                employeeTableViewController.managedObjectContext = aContext
            }
            let share: CKShare? = synchronizer?.share(for: company!)
            employeeTableViewController.canWrite = share?.currentUserParticipant?.permission == .readWrite
        }
    }

    // MARK: - NSFetchedResultsControllerDelegate
    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.beginUpdates()
    }
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange sectionInfo: NSFetchedResultsSectionInfo, atSectionIndex sectionIndex: Int, for type: NSFetchedResultsChangeType) {
        switch type {
        case .insert:
            tableView.insertSections(NSIndexSet(index: sectionIndex) as IndexSet, with: .fade)
        case .delete:
            tableView.deleteSections(NSIndexSet(index: sectionIndex) as IndexSet, with: .fade)
        default:
            break
        }
    }
    
    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
        let tableView: UITableView? = self.tableView
        switch type {
        case .insert:
            tableView?.insertRows(at: [newIndexPath!], with: .fade)
        case .delete:
            tableView?.deleteRows(at: [indexPath!], with: .fade)
        case .update:
            if let aPath = indexPath {
                configureCell(tableView?.cellForRow(at: aPath) as? QSCompanySwiftTableViewCell, at: indexPath)
            }
        case .move:
            tableView?.deleteRows(at: [indexPath!], with: .fade)
            tableView?.insertRows(at: [newIndexPath!], with: .fade)
        }
    }

    
    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        tableView.endUpdates()
    }
    
    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return .delete
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let company = object(at: indexPath) as? QSCompany
        performSegue(withIdentifier: "showEmployees", sender: company)
    }
    
// MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        
        if let frc = fetchedResultsController {
            return frc.fetchedResultsControllers!.count
        }
        return 0
    }
    
    override func tableView(_ table: UITableView, numberOfRowsInSection section: Int) -> Int {
        

        if let frc = fetchedResultsController {
            return frc.fetchedResultsControllers[section].fetchedObjects?.count ?? 0
        }
        return 0
        
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell") as? QSCompanySwiftTableViewCell
        configureCell(cell, at: indexPath)
        if let aCell = cell {
            return aCell
        }
        return UITableViewCell()
    }
    
    func configureCell(_ cell: QSCompanySwiftTableViewCell?, at indexPath: IndexPath?) {
        var company: QSCompany? = nil
        if let aPath = indexPath {
            company = object(at: aPath)
        }
        cell?.nameLabel.text = company?.name
        cell?.sharingButton.setTitle("Shared with me", for: .normal)
        cell?.shareButtonAction = {
            self.share(company!)
        }
    }
    
    func object(at indexPath: IndexPath) -> QSCompany {
        if let aRow = fetchedResultsController?.fetchedResultsControllers[indexPath.section].fetchedObjects![indexPath.row] {
            return aRow as! QSCompany
        }
        return QSCompany()
    }

    
    
// MARK: - Actions
    
//    func subscribe() {
//
//        let artistRecordID = CKRecordID(recordName: "Mei Chen")
//        let predicate = NSPredicate(format: "artist = %@", artistRecordID)
//
//        let subscription = CKSubscription(recordType: "Artwork", predicate: predicate, options: .firesOnRecordCreation)
//        let DB = CKContainer.default()
//
//        let CKSubscription = CKRecordZoneSubscription(zoneID: artistRecordID.zoneID)
//        var notificationInfo = CKNotificationInfo()
//        notificationInfo.alertLocalizationKey = "New artwork by your favorite artist."
//        notificationInfo.shouldBadge = true
//
//
//        subscription.notificationInfo = notificationInfo
//
//
//        var publicDatabase: CKDatabase? = CKContainer.default().publicCloudDatabase
//        publicDatabase?.save(subscription, completionHandler: { (subsription, error) in
//            print(error)
//        })
//
//    }
    
    func subscribeChanges() {
        let container = CKContainer.default()
        let subscription = CKDatabaseSubscription(subscriptionID: "test")
        let notificationInfo = CKNotificationInfo()
        notificationInfo.shouldSendContentAvailable = true
        subscription.notificationInfo = notificationInfo
        
        let operation = CKModifySubscriptionsOperation(subscriptionsToSave: [subscription], subscriptionIDsToDelete: [])
        operation.modifySubscriptionsCompletionBlock = { savedSubscriptions, deletedSubscriptionIDs, operationError in
            if operationError != nil {
                print(operationError)
                return
            } else {
                print("Subscribed")
            }
        }
        
//        container.privateCloudDatabase.add(operation)
        container.sharedCloudDatabase.add(operation)
    }
    
    @IBAction func didTapSynchronize(_ sender: Any) {
        synchronize(withCompletion: { _ in })
    }
    
    func synchronize(withCompletion completion: ((_ error: Error?) -> Void)? = nil) {
        self.showLoading(true)
        synchronizer?.synchronize(completion: { error in
            self.showLoading(false)
            if error != nil {
                var alertController: UIAlertController? = nil
                if let anError = error {
                    alertController = UIAlertController(title: "Error", message: "Error: \(anError)", preferredStyle: .alert)
                }
                alertController?.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                if let aController = alertController {
                    self.present(aController, animated: true)
                }
            }
            
            if (completion != nil) {
                completion!(error)
            }
        })
    }
    
    func share(_ company: QSCompany) {
        sharingCompany = company
        showLoading(true)
        synchronizer?.synchronize(completion: { error in
            self.showLoading(false)
            if error == nil {
                var sharingController: UICloudSharingController?
                let share: CKShare? = self.synchronizer?.share(for: company)
                let container = CKContainer(identifier: self.synchronizer?.containerIdentifier ?? "")
                if share != nil {
                    if let aShare = share {
                        sharingController = UICloudSharingController(share: aShare, container: container)
                    }
                } else {
                    sharingController = UICloudSharingController(preparationHandler: { controller, preparationCompletionHandler in
                        self.synchronizer!.share(object: company, publicPermission: CKShareParticipantPermission.readOnly, participants: []) { share, error in
                            preparationCompletionHandler(share, container, error)
                        }
                    })
                }
                sharingController?.availablePermissions = [.allowPublic, .allowReadOnly, .allowReadWrite]
                sharingController?.delegate = self
                if let aController = sharingController {
                    self.present(aController, animated: true)
                }
            }
        })
    }

    // MARK: - UICloudSharingControllerDelegate
    func itemTitle(for csc: UICloudSharingController) -> String? {
        return sharingCompany?.name
    }
    
    func cloudSharingController(_ csc: UICloudSharingController, failedToSaveShareWithError error: Error) {
        print("failed to save share")
    }

    func cloudSharingControllerDidSaveShare(_ csc: UICloudSharingController) {
        synchronizer?.saveShare(csc.share!, for: sharingCompany)
        synchronizer?.synchronize(completion: nil)
    }

    func cloudSharingControllerDidStopSharing(_ csc: UICloudSharingController) {
        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + Double(Int64(2 * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC), execute: {
            self.synchronize(withCompletion: nil)
        })
    }
    
    func multiFetchedResultsControllerDidChangeControllers(_ controller: QSCoreDataMultiFetchedResultsController?) {
        tableView.reloadData()
    }

    // MARK: - Loading
    func showLoading(_ loading: Bool) {
        if loading {
            syncButton.isHidden = true
            loadingView?.isHidden = false
            indicatorView.startAnimating()
        } else {
            syncButton.isHidden = false
            loadingView?.isHidden = true
            indicatorView.stopAnimating()
        }
    }

}
